executors/cmdexec/volume.go

```diff
@@ -276,6 +276,42 @@ func (s *CmdExecutor) VolumeInfo(host string, volume string) (*executors.Volume,
 	return &volumeInfo.VolInfo.Volumes.VolumeList[0], nil
 }
 
+func (s *CmdExecutor) VolumeStatusDetailed(host string, volume string) (*executors.VolumeDetail, error) {
+	godbc.Require(volume != "")
+	godbc.Require(host != "")
+
+	type CliOutput struct {
+		OpRet     int                 `xml:"opRet"`
+		OpErrno   int                 `xml:"opErrno"`
+		OpErrStr  string              `xml:"opErrstr"`
+		VolStatus executors.VolStatus `xml:"volStatus"`
+	}
+
+	command := []string{
+		fmt.Sprintf("gluster --mode=script volume status %v detail --xml", volume),
+	}
+
+	//Get the xml output of volume info
+	output, err := s.RemoteExecutor.RemoteCommandExecute(host, command, 10)
+	if err != nil {
+		return nil, fmt.Errorf("Unable to get detailed volume status of volume name: %v", volume)
+	}
+	var volumeDetails CliOutput
+	err = xml.Unmarshal([]byte(output[0]), &volumeDetails)
+	if err != nil {
+		return nil, fmt.Errorf("Unable to determine detailed volume status of volume name: %v", volume)
+	}
+	logger.Debug("%+v\n", volumeDetails)
+
+	volumeStatus := &volumeDetails.VolStatus.Volumes.VolumeDetailList[0]
```

Comment:
In case `volumeDetails.OpRet != 0`, it is likely that the `volumeDetails` struct is not complete. Instead of fetching the `volumeStatus.VolumeName` from the XML, you can just use the `volume` argument passed to the function.

This should make the code a little more safe in case of unexpected errors from the `gluster` command.

Reply:
Done. Now using heketi volume ID instead of gluster name.